### ISSUE TITLE
Usage of scalar-leaf trees with multi-output objectives

### ISSUE THREAD

**Issue by david-cortes** [dmlc/xgboost#12486]
In https://github.com/dmlc/xgboost/issues/12278 , it was mentioned:
> In one tree-per-class mode, one idea would be to recompute the gradients after each tree, instead of after each group of trees. This turns it into a coordinate descent algorithm which is much better at dealing with curvature.

Following this example:
https://xgboost.readthedocs.io/en/stable/tutorials/advanced_custom_obj.html

If setting parameters `num_boost_round=1` and `base_score=0` without any base_score, then calling something like:
```python
booster.inplace_predict(X, predict_type="margin")
```

I would expect from that comment and from the documentation to get a single tree making predictions for a single column of the output, but `.predict()` produces an output where each column has a non-zero value, even though the custom objective is called only once (verified by putting prints on it) and there is a single tree, as verified by:

```python
import json
trees = json.loads(booster.save_raw(raw_format="json"))["learner"]["gradient_booster"]["model"]["trees"]
len(trees)
```

What's happening there?

**Comment by trivialfis** [comment]
"one idea would be"

We don't use that idea. It would be, if we use the idea.

The objective is called once for all targets.

**Comment by david-cortes** [comment]
> "one idea would be"
> 
> We don't use that idea. It would be, if we use the idea.
> 
> The objective is called once for all targets.

Still: with `multi_strategy=one_output_per_tree`, shouldn't a single tree make predictions for a single class?

**Comment by trivialfis** [comment]
> Still: with multi_strategy=one_output_per_tree, shouldn't a single tree make predictions for a single class?

For one output per tree, each boosting iteration creates K trees, with K equal to the number of classes. With a multi-output tree (vector leaf), each boosting round adds a single tree for all classes.

The objective function is called once per boosting round to provide gradient and Hessian for all targets/classes.

**Comment by david-cortes** [comment]
> > Still: with multi_strategy=one_output_per_tree, shouldn't a single tree make predictions for a single class?
> 
> For one output per tree, each boosting iteration creates K trees, with K equal to the number of classes. With a multi-output tree (vector leaf), each boosting round adds a single tree for all classes.
> 
> The objective function is called once per boosting round to provide gradient and Hessian for all targets/classes.

But then what is the json exporter giving? It shows a single tree with regular nodes that have left/right branches, leaf values, and so on. If following that tree for a given observation, it would end up at one leaf with one value.

**Comment by trivialfis** [comment]
I'm a bit confused by the question, could you please share a script that I can run and point out which output doesn't match your expectation?

**Comment by david-cortes** [comment]
Sorry about earlier comment, tried the latest version now and the json does end up having one tree per column of 'y'.

> The objective is called once for all targets.

Sounds like that aspect could be improved. What would be the issue with calculating gradients/hessians for one target at a time after each single-target tree? (as described at the beginning)

**Comment by trivialfis** [comment]
For independent outputs, I don't see obvious issue, at least not on top of my head. Might even be preferable. One can create a Python prototype to experiment and see whether it improves the held out loss. The booster class can be updated without using the `train` function.

**Comment by RAMitchell** [comment]
I think you would have a quadratic complexity with number of outputs - you need to compute gradients for n_targets, n_targets times. I'm confident it would improve training outcomes for multi-class but that would be slow for large n_targets. 

We have some nice alternatives coming in vector leaf so I would rather focus on that.

**Comment by david-cortes** [comment]
> I think you would have a quadratic complexity with number of outputs - you need to compute gradients for n_targets, n_targets times. I'm confident it would improve training outcomes for multi-class but that would be slow for large n_targets.
> 
> We have some nice alternatives coming in vector leaf so I would rather focus on that.

No, it would only need to compute gradients/hessians for one target each time, holding predictions for the others constant. It would require introducing new calculations in the objectives to do it efficiently, but the complexity would be the same.

**Comment by trivialfis** [comment]
Please distinguish between dependent outputs and independent outputs.

**Comment by david-cortes** [comment]
> Please distinguish between dependent outputs and independent outputs.

What do you mean?

**Comment by trivialfis** [comment]
The question depends on when calculating one target gradient needs to calculate gradient from other targets, like the softmax objective.

**Comment by david-cortes** [comment]
> The question depends on when calculating one target gradient needs to calculate gradient from other targets, like the softmax objective.

For a given observation, assuming $k$ classes, the Hessian for class $c$ is given as:
$p_c \times (1 - p_c)$

Where
$p_c = \frac{ \exp(x_c) }{ \sum_{i=1}^k \exp(x_i) }$

For that, as long as you have a running version of $\sum_{i=1}^k \exp(x_i)$, it's only $O(\text{nrows})$ operations to calculate row-wise hessians for a single class.

After fitting a tree for a single class, then the running sum-exp can be updated with $O(\text{nrows})$ operations too, and the process then repeated for each class.

To make it more precise, one might want to keep also something along the lines of $\max_{i=1}^k \exp(x_i)$ from previous rounds and calculate instead the numbers as:
$p_c = \frac{ \exp(x_c - \text{max}) }{ \sum_{i=1}^k \exp(x_i - \text{max}) }$

And those can be updated only after a full cycle to keep the procedure at $O(\text{nrows} \times k)$ operations per cycle. Although in practice $k$ is small enough that I don't think it'd matter if it's $O(\text{nrows} \times k^2)$ operations per cycle.
